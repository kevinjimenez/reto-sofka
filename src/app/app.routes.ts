import { Routes } from '@angular/router';
import { LayoutComponent } from './layout/layout.component';

export const routes: Routes = [
  {
    path: 'products',
    component: LayoutComponent,
    children: [
      {
        path: '',
        loadChildren: () =>
          import('./pages/products/products.routes').then((m) => m.routes),
      },
      {
        path: '',
        redirectTo: '',
        pathMatch: 'full',
      },
    ],
  },
  {
    path: '',
    redirectTo: 'products',
    pathMatch: 'full',
  },
  // {
  //   path: '**',
  //   loadComponent: () =>
  //     import('./pages/not-found/not-found.component').then(m => m.default),
  // },
];
