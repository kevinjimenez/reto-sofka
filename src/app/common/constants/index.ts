export { CONTEXT_MENU_OPTIONS } from './context-menu-options';
export { PRODUCT_TABLE_HEADERS } from './table-headers';
export { ERROR_INPUT_MESSAGE } from './error-input-message';
