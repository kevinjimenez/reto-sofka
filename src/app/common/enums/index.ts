export { MENU_OPTIONS } from './menu-options';
