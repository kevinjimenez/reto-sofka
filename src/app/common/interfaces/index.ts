export { Product } from './product.interface';
export { DefaultResponse } from './default-response.interface';
export { TableHeaderInterface } from './table-header.interface';
export { ErrorInputInterface } from './error-input.interface';
