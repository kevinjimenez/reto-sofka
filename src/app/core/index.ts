export { spinnerInterceptor } from './interceptors/spinner.interceptor';
export { productsResolver } from './resolvers/products.resolver';
export { ProductsService } from './services/products.service';
export { SpinnerService } from './services/spinner.service';
