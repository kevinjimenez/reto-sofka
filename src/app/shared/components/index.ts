export { ContextMenuOption } from './context-menu/interfaces/context-menu-option.interface';
export { ContextMenuOptionSelected } from './context-menu/interfaces/context-menu-option-selected.interface';
export { ContextMenuComponent } from './context-menu/context-menu.component';
export { IconButtonComponent } from './icon-button/icon-button.component';
export { AlertCircleComponent } from './icons/alert-circle/alert-circle.component';
export { OverlayComponent } from './overlay/overlay.component';
export { SpinnerComponent } from './spinner/spinner.component';
export { ButtonComponent } from './button/button.component';
export { InputComponent } from './input/input.component';
export { LogoComponent } from './logo/logo.component';
export { ToastComponent } from './toast/toast.component';
export { TooltipComponent } from './tooltip/tooltip.component';
