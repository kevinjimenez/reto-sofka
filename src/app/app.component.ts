import { Component } from '@angular/core';
import { RouterOutlet } from '@angular/router';
import { SpinnerComponent } from './shared/components';

@Component({
	selector: 'app-root',
	standalone: true,
	imports: [RouterOutlet, SpinnerComponent],
	templateUrl: './app.component.html',
	styleUrl: './app.component.css'
})
export class AppComponent {}
