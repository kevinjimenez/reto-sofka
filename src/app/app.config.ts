import { ApplicationConfig, provideZoneChangeDetection } from '@angular/core';
import { provideRouter, withComponentInputBinding, withHashLocation } from '@angular/router';

import { provideHttpClient, withInterceptors } from '@angular/common/http';
import { routes } from './app.routes';
import { spinnerInterceptor } from './core';

export const appConfig: ApplicationConfig = {
	providers: [
		provideZoneChangeDetection({ eventCoalescing: true }),
		provideHttpClient(withInterceptors([spinnerInterceptor])),
		provideRouter(
			routes,
			withHashLocation(),
			withComponentInputBinding()
			// withPreloading(PreloadAllModules)
		)
	]
};
