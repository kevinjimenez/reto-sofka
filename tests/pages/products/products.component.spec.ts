import { DebugElement } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { of } from 'rxjs';
import { Product } from '../../../src/app/common/interfaces';
import { ProductsService } from '../../../src/app/core';
import { TableComponent } from '../../../src/app/pages/products/components/table/table.component';
import { ProductsComponent } from '../../../src/app/pages/products/products.component';
import { ButtonComponent, InputComponent } from '../../../src/app/shared/components';
import { By } from '@angular/platform-browser';

const productsMock: Product[] = [
	{
		id: '1',
		name: 'Product 1',
		description: 'Description 1',
		logo: 'logo1.png',
		date_release: '2024-01-01',
		date_revision: '2024-01-01'
	},
	{
		id: '2',
		name: 'Product 2',
		description: 'Description 2',
		logo: 'logo2.png',
		date_release: '2024-01-01',
		date_revision: '2024-01-01'
	}
];

describe('ProductsComponent', () => {
	let component: ProductsComponent;
	let fixture: ComponentFixture<ProductsComponent>;
	let compiled: DebugElement;
	let routerMock = {
		navigate: jest.fn()
	};
	let activatedRouteMock = {
		data: of({
			products: {
				data: productsMock
			}
		})
	};
	let productsServiceMock = {};

	beforeEach(async () => {
		await TestBed.configureTestingModule({
			imports: [
				ProductsComponent,
				TableComponent,
				InputComponent,
				ButtonComponent,
				ReactiveFormsModule
			],
			providers: [
				{ provide: ProductsService, useValue: productsServiceMock },
				{ provide: Router, useValue: routerMock },
				{ provide: ActivatedRoute, useValue: activatedRouteMock }
			]
		}).compileComponents();
	});

	beforeEach(() => {
		fixture = TestBed.createComponent(ProductsComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();
		compiled = fixture.debugElement;
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('should match snapshot', () => {
		expect(compiled).toMatchSnapshot();
	});

	it('should initialize products', () => {
		const originalProducts = component.originalProducts();
		expect(originalProducts).toEqual(productsMock);

		const cloneProducts = component.cloneProducts();
		expect(cloneProducts).toEqual(originalProducts.slice(0, component.take()));
	});

	it('should navigate to Form when "Agregar" button is clicked', () => {
		const onNewProductSpy = jest.spyOn(component, 'onNewProduct');
		const buttonComponent = compiled.query(By.directive(ButtonComponent));
		buttonComponent.triggerEventHandler('onClick', null);

		expect(onNewProductSpy).toHaveBeenCalled();
		expect(routerMock.navigate).toHaveBeenCalledWith(['products', 'product']);
	});

	it('should update cloneProducts on search input', (done) => {
		component.fieldSearch.setValue('Product 1');
		setTimeout(() => {
			expect(component.cloneProducts()).toEqual(productsMock.slice(0, 1));
			done();
		}, 600);
	});

	it('should reset cloneProducts when search input is cleared', (done) => {
		component.fieldSearch.setValue('Product 1');
		setTimeout(() => {
			component.fieldSearch.setValue('');
			setTimeout(() => {
				expect(component.cloneProducts()).toEqual(
					component.originalProducts().slice(0, component.take())
				);
				done();
			}, 600);
		}, 600);
	});

	it('should call onRemoveItem and remove a product from the list', () => {
		const initialLength = component.cloneProducts().length;
		const productToRemove = '1';
		component.onRemoveItem(productToRemove);
		expect(component.cloneProducts().length).toBe(initialLength - 1);
	});

	it('should call onViewItems and update visible products based on "take" value', () => {
		component.onViewItems(1);
		expect(component.cloneProducts().length).toBe(1);
		expect(component.cloneProducts()).toEqual(productsMock.slice(0, 1));
	});
});
