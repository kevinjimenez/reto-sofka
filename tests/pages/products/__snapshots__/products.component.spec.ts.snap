// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsComponent should match snapshot 1`] = `
DebugElement {
  "nativeNode": <div
    id="root1"
  >
    <main
      class="container"
    >
      <header
        class="flex justify-between items-center flex-row gap-1 my-1"
      >
        <app-input>
          <section
            class="button-container"
          >
            
            <input
              class="valid ng-untouched ng-pristine ng-valid"
              id="undefined"
              placeholder="Search"
              type="text"
            />
            
          </section>
        </app-input>
        <app-button
          color="second-brand"
          label="Agregar"
        >
          <button
            class="secundary"
            type="button"
          >
             Agregar

          </button>
        </app-button>
      </header>
      <app-table>
        <app-toast>
          
        </app-toast>
        <section
          class="table-container"
        >
          <div
            class="table-wrapper"
          >
            <table>
              <thead>
                <tr>
                  <th>
                    <section
                      class="flex items-center w-full gap-1/2 cursor-default"
                    >
                       Logo 
                      
                    </section>
                  </th>
                  <th>
                    <section
                      class="flex items-center w-full gap-1/2 cursor-default"
                    >
                       Nombre del producto 
                      
                    </section>
                  </th>
                  <th>
                    <section
                      class="flex items-center w-full gap-1/2 cursor-default"
                    >
                       Descripción 
                      <span
                        class="tooltip-container"
                      >
                        <app-alert-circle
                          class="cursor-pointer"
                        >
                          <svg
                            class="primary"
                            fill="currentColor"
                            height="24"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M0 0h24v24H0z"
                              fill="none"
                              stroke="none"
                            />
                            <path
                              d="M12 2c5.523 0 10 4.477 10 10a10 10 0 0 1 -19.995 .324l-.005 -.324l.004 -.28c.148 -5.393 4.566 -9.72 9.996 -9.72zm.01 13l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-.01 -8a1 1 0 0 0 -.993 .883l-.007 .117v4l.007 .117a1 1 0 0 0 1.986 0l.007 -.117v-4l-.007 -.117a1 1 0 0 0 -.993 -.883z"
                            />
                          </svg>
                        </app-alert-circle>
                        <app-tooltip>
                          <main
                            class="tooltip-content"
                          >
                             Descripción

                          </main>
                        </app-tooltip>
                      </span>
                      
                    </section>
                  </th>
                  <th>
                    <section
                      class="flex items-center w-full gap-1/2 cursor-default"
                    >
                       Fecha de liberación 
                      <span
                        class="tooltip-container"
                      >
                        <app-alert-circle
                          class="cursor-pointer"
                        >
                          <svg
                            class="primary"
                            fill="currentColor"
                            height="24"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M0 0h24v24H0z"
                              fill="none"
                              stroke="none"
                            />
                            <path
                              d="M12 2c5.523 0 10 4.477 10 10a10 10 0 0 1 -19.995 .324l-.005 -.324l.004 -.28c.148 -5.393 4.566 -9.72 9.996 -9.72zm.01 13l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-.01 -8a1 1 0 0 0 -.993 .883l-.007 .117v4l.007 .117a1 1 0 0 0 1.986 0l.007 -.117v-4l-.007 -.117a1 1 0 0 0 -.993 -.883z"
                            />
                          </svg>
                        </app-alert-circle>
                        <app-tooltip>
                          <main
                            class="tooltip-content"
                          >
                             Fecha de liberación

                          </main>
                        </app-tooltip>
                      </span>
                      
                    </section>
                  </th>
                  <th>
                    <section
                      class="flex items-center w-full gap-1/2 cursor-default"
                    >
                       Fecha de reestructuración 
                      <span
                        class="tooltip-container"
                      >
                        <app-alert-circle
                          class="cursor-pointer"
                        >
                          <svg
                            class="primary"
                            fill="currentColor"
                            height="24"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M0 0h24v24H0z"
                              fill="none"
                              stroke="none"
                            />
                            <path
                              d="M12 2c5.523 0 10 4.477 10 10a10 10 0 0 1 -19.995 .324l-.005 -.324l.004 -.28c.148 -5.393 4.566 -9.72 9.996 -9.72zm.01 13l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-.01 -8a1 1 0 0 0 -.993 .883l-.007 .117v4l.007 .117a1 1 0 0 0 1.986 0l.007 -.117v-4l-.007 -.117a1 1 0 0 0 -.993 -.883z"
                            />
                          </svg>
                        </app-alert-circle>
                        <app-tooltip>
                          <main
                            class="tooltip-content"
                          >
                             Fecha de reestructuración

                          </main>
                        </app-tooltip>
                      </span>
                      
                    </section>
                  </th>
                  <th>
                    <section
                      class="flex items-center w-full gap-1/2 cursor-default"
                    >
                        
                      
                    </section>
                  </th>
                  
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    <span
                      class="flex justify-center items-center w-full"
                    >
                      <app-logo>
                        <img
                          alt="__logo__"
                          class="logo"
                          onerror=""
                          src="logo1.png"
                        />
                        
                        
                      </app-logo>
                    </span>
                  </td>
                  <td>
                    Product 1
                  </td>
                  <td>
                    Description 1
                  </td>
                  <td>
                    2024-01-01
                  </td>
                  <td>
                    2024-01-01
                  </td>
                  <td>
                    <app-icon-button
                      icon="dots-vertical.svg"
                    >
                      <button
                        class="container"
                        type="button"
                      >
                        <img
                          alt="dots-vertical.svg"
                          src="assets/icons/dots-vertical.svg"
                        />
                      </button>
                    </app-icon-button>
                    <app-context-menu>
                      
                    </app-context-menu>
                  </td>
                </tr>
                <tr>
                  <td>
                    <span
                      class="flex justify-center items-center w-full"
                    >
                      <app-logo>
                        <img
                          alt="__logo__"
                          class="logo"
                          onerror=""
                          src="logo2.png"
                        />
                        
                        
                      </app-logo>
                    </span>
                  </td>
                  <td>
                    Product 2
                  </td>
                  <td>
                    Description 2
                  </td>
                  <td>
                    2024-01-01
                  </td>
                  <td>
                    2024-01-01
                  </td>
                  <td>
                    <app-icon-button
                      icon="dots-vertical.svg"
                    >
                      <button
                        class="container"
                        type="button"
                      >
                        <img
                          alt="dots-vertical.svg"
                          src="assets/icons/dots-vertical.svg"
                        />
                      </button>
                    </app-icon-button>
                    <app-context-menu>
                      
                    </app-context-menu>
                  </td>
                </tr>
                
                
                
              </tbody>
            </table>
          </div>
          <footer
            class="table-footer"
          >
            <span>
              2 Resultados
            </span>
            <select
              class="ng-untouched ng-pristine ng-valid"
            >
              <option
                value="5"
              >
                5
              </option>
              <option
                value="10"
              >
                10
              </option>
              <option
                value="15"
              >
                15
              </option>
            </select>
          </footer>
        </section>
        <app-overlay>
          
        </app-overlay>
      </app-table>
    </main>
  </div>,
}
`;
